services/headless-lms/src/models/pages.rs

```diff
@@ -447,14 +447,30 @@ async fn upsert_exercises_and_exercise_tasks(
     .fetch_all(&mut *conn)
     .await?;
 
-    let existing_exercise_task_ids = sqlx::query!(
-            r#"
-    SELECT exercise_tasks.id from exercise_tasks JOIN exercises e ON (e.id = exercise_tasks.exercise_id) WHERE page_id = $1
+    // let exercise_infos_by_type = get_all_exercise_services_by_type(conn).await?;
+    let existing_exercise_tasks = sqlx::query_as!(
+        ExerciseTaskIdAndSpec,
+        r#"
+SELECT et.id,
+  et.private_spec,
+  et.public_spec
+from exercise_tasks et
+  JOIN exercises e ON (e.id = et.exercise_id)
+WHERE page_id = $1
         "#,
-            page.id
-        )
-        .fetch_all(&mut *conn)
-        .await?;
+        page.id
+    )
+    .fetch_all(&mut *conn)
+    .await?;
+
+    // For generating public specs for exercises.
+    let client = reqwest::Client::new();
+    let public_spec_urls_by_exercise_type = get_all_exercise_services_by_type(conn)
+        .await?
+        .into_iter()
+        .map(|(key, (service, info))| Ok((key, get_internal_public_spec_url(&service, &info)?)))
```

Comment:
Use only necessary services here because someone might add a invalid exercise service to the db that would crash this always